deployment/WFCORE-3522_Deployment content validation.adoc

```diff
@@ -0,0 +1,67 @@
+= Title
+:author:            David Lloyd
+:email:             david.lloyd@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          deployment
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+We need to move to child-first class and resource resolution in deployments.
+The first step must be validation of deployment contents and dependencies.
+This entails detecting deployment artifacts which have packages or resources which conflict with container packages or resources.
+
+Once this proposal is implemented, we can move to changing dependency order, which will come in a follow-up proposal.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-3522[WFCORE-3522]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/SWARM-1691[SWARM-1691]
+* https://issues.jboss.org/browse/WFLY-9828[WFLY-9828]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mjurc@redhat.com[Michal Jurc]
+
+=== Affected Projects or Components
+
+* WildFly Core
+* All deployment-related subsystems
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* The deployer must provide a way for subsystems to register packages and resources that are indicative of invalid deployment content.
+This likely entails introducing a multi-valued capability which is consumed by the deployment system.
```

Comment:
Is the integration of subsystems with the new capability a concern of this feature? Or is this feature just about the implementation of capability itself, and subsystems using it will be a later step?

Reply:
That was my initial plan.  I don't really know if it's realistic to expect 100% coverage at first (maybe it is) but my thought was that if we don't put it in subsystems right away, the feature would not be useful.

That said, I think that any number of subsystems could be verified using just one test where one resource is generated into each forbidden package and verified in a loop, or something like that.  Then it's just a question of ensuring that we found all the packages we want to forbid, which I think would have to be a manual verification.